### PR TITLE
Documentation: update language reference

### LINKED ISSUE #1627: Update language reference

### PR DESCRIPTION
**Pull request by lukaszcz**
* Closes #1627

### CODE REVIEW

File: docs/org/reference/language/functions.org [hunk 1]
```diff
@@ -47,3 +47,41 @@ odd (suc n) := even n;
 even zero := true;
 even (suc n) := odd n;
 #+end_example
+
+** Anonymous functions
+
+Anonymous functions, or /lambdas/, are introduced with the syntax:
+
+#+begin_example
+\{pat1 .. patN_1 := clause1;
```
Comment (janmasrovira) [inline docs/org/reference/language/functions.org:56]: I think we should enourage, or at least mention, that lambdas can be written using the unicode symbol

File: docs/org/reference/language/functions.org [hunk 2]
```diff
@@ -47,3 +47,41 @@ odd (suc n) := even n;
...
+  ..
+  pat1 .. patN_M := clauseM}
```
Comment (janmasrovira) [inline docs/org/reference/language/functions.org:58]: lambda clauses are separated by `|` rather than `;`

File: docs/org/reference/language/functions.org [hunk 3]
```diff
@@ -47,3 +47,41 @@ odd (suc n) := even n;
...
+  zero := true;
+  (suc n) := odd n;
+};
+#+end_example
+
+** Short definitions
+
+If a function has only one clause, then its definition can written in one line, with the body immediately following the signature:
+#+begin_example
+multiplyByTwo : Nat -> Nat := \{n := 2 * n};
```
Comment (janmasrovira) [inline docs/org/reference/language/functions.org:86]: and here we are using ascii arrow. We should be consistent

File: docs/org/reference/language/control.org [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+#+begin_example
+case value
+| pat1 := branch1
+..
+| patN := branchN
+#+end_example
+
+For example, one can evaluate the following expression in the REPL:
+#+begin_example
+Stdlib.Prelude> case (1, 2) | (suc _, zero) := 0 | (suc _, suc x) := x | _ := 19
```
[the remark this answers is not in the document]
Comment (lukaszcz) [reply docs/org/reference/language/control.org:16]: Done

File: docs/org/reference/language/functions.org (hunk 1 above)
Comment (lukaszcz) [reply docs/org/reference/language/functions.org:56]: Done

File: docs/org/reference/language/functions.org (hunk 2 above)
Comment (lukaszcz) [reply docs/org/reference/language/functions.org:58]: Done

File: docs/org/reference/language/functions.org [hunk 5]
```diff
@@ -47,3 +47,41 @@ odd (suc n) := even n;
...
+#+end_example
+
+An anonymous function just lists all clauses of a function without
+naming it. Any function declaration can be converted to use anonymous
+functions:
+
+#+begin_example
+open import Stdlib.Prelude;
+
+odd : Nat → Bool;
```
[the remark this answers is not in the document]
Comment (lukaszcz) [reply docs/org/reference/language/functions.org:68]: Done. I changed to unicode arrows.

File: docs/org/reference/language/functions.org (hunk 3 above)
Comment (lukaszcz) [reply docs/org/reference/language/functions.org:86]: Done

File: docs/org/reference/language/functions.org (hunk 5 above)
[the remark this answers is not in the document]
Comment (lukaszcz) [reply docs/org/reference/language/functions.org:70]: Well, actually in other files we use ASCII arrows, so we should change to ASCII by default.

File: docs/org/reference/language/control.org [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+#+begin_example
+case value
+| pat1 := branch1
+..
+| patN := branchN
+#+end_example
+
+For example, one can evaluate the following expression in the REPL:
+#+begin_example
+Stdlib.Prelude> case 1, 2 | suc _, zero := 0 | suc _, suc x := x | _ := 19
```
Comment (janmasrovira) [inline docs/org/reference/language/control.org:16]: I feel like this will be a bit hard to parse for a newcomer. I would prefer to see a simpler `case` expression in the repl or write it multiline outside of the repl (or both)

Comment (lukaszcz) [reply docs/org/reference/language/control.org]: Okay, I simplified the case expression by removing the pair

File: docs/org/reference/language/functions.org [hunk 7]
```diff
@@ -37,13 +37,52 @@ Function declarations can depend on each other recursively. In the following exa
...
+
+odd := \{
+  zero := false;
+  (suc n) := even n;
+};
+
+even := \{
+  zero := true;
+  (suc n) := odd n;
+};
```
Comment (jonaprieto) [inline docs/org/reference/language/functions.org:80]: Replace this by:

```
  odd := \ {
      | zero := false
      | (suc n) := even n
    };

  even := \ {
      | zero := true
      | (suc n) := odd n
    };
```

Comment (lukaszcz) [reply docs/org/reference/language/functions.org:80]: Right, forgot about the pipes again

File: docs/org/reference/language/functions.org [hunk 8]
```diff
@@ -37,13 +37,52 @@ Function declarations can depend on each other recursively. In the following exa
...
+#+end_example
+
+** Short definitions
+
+If a function has only one clause, then its definition can written in one line, with the body immediately following the signature:
```
Comment (jonaprieto) [inline docs/org/reference/language/functions.org:85]: Please reformulate this sentence. Also, consider that if I have this function with two clauses:

```
  giveMeTheTruth : Nat -> Bool := λ {
      | zero := true
      | (suc n) := false
    };
```
I can also write in one line `giveMeTheTruth : Nat -> Bool := λ { | zero := true | (suc n):= false };`

File: docs/org/reference/language/functions.org (hunk 7 above)
Comment (jonaprieto) [reply docs/org/reference/language/functions.org:80]: It's fine; we'll address the main issue here: 
- #1839

File: docs/org/reference/language/functions.org (hunk 8 above)
Comment (lukaszcz) [reply docs/org/reference/language/functions.org:85]: Well, this function has one clause whose body is a lambda with two clauses. How else can express the fact that there is only one clause in the function *definition*? I just removed any mention of clauses, if it's confusing.

File: docs/org/reference/language/functions.org (hunk 7 above)
Comment (lukaszcz) [reply docs/org/reference/language/functions.org]: Done

File: docs/org/reference/language/functions.org (hunk 8 above)
Comment (jonaprieto) [reply docs/org/reference/language/functions.org]: Yes, it was a bit confusing.
You still have a typo. It should be "can be written"

Comment (lukaszcz) [reply docs/org/reference/language/functions.org]: I already fixed the typo.

Comment (jonaprieto) [reply docs/org/reference/language/functions.org]: Thanks. I see it now.

**Review by jonaprieto — APPROVED** [review]
